gh-87281: Improve documentation for locale.setlocale() and locale.getlocale() (#137313)
python/cpython: Doc/library/locale.rst

   If *locale* is given and not ``None``, :func:`setlocale` modifies the locale
   setting for the *category*. The available categories are listed in the data
   description below. *locale* may be a string, or a pair,
   language code and encoding. If it is a pair, it is converted to a locale
   name using the locale aliasing engine. An empty string specifies the user's
   default settings. If the modification of the locale fails, the exception
   :exc:`Error` is raised. If successful, the new locale setting is returned.

   The format of the *locale* and the language code strings is platform

>    depended, but the forms ``language[_territory][.encoding][@modifier]``

Reviewer: ```suggestion
   dependent, but the forms ``language[_territory][.encoding][@modifier]``
```

>    and ``language[_territory]`` respectively are typically accepted on all

Reviewer: ```suggestion
   and ``language[_territory]`` are typically accepted on all
```
I'm not sure what this is referring to?

Author: It refers to the locale and the language code strings.

Author: If this is not clear, we can only specify format for locale, and then say that the language code has the same format, but without encoding and modifier (support for modifiers will be added in #137253).

Reviewer 2: I'd write: "the form ``language[_territory][.encoding][@modifier]`` is typically accepted".

The second version you mention is already included in this.

   ``'LANG'``.  The GNU gettext search path contains ``'LC_ALL'``,
   ``'LC_CTYPE'``, ``'LANG'`` and ``'LANGUAGE'``, in that order.

>    The format of the language code is platform depended, but on Posix

Reviewer: This function is (going to be) removed in 3.15, there is no need to update its docs.

Author: But it will not be removed in 3.13 and 3.14.

   platforms it usually looks like ``language[_territory]``.

>    The language code and encoding may be ``None`` if their values cannot be

Reviewer 2: Please readd the mention of the RF: "looks like ``language[_territory]`` (see :rfc:`1766`)."

Author: RFC 1766 does not look relevant.

* RFC 1766 was obsoleted by RFC 3066 which was obsoleted by RFC 4646 and RFC 4646 (aka BCP 47). BCP 47 is the current industrial standard. A limited subset of BCP 47 is supported on Windows, but not on Posix platforms.
* All these standards use a hyphen as a separator, not an underscore. Names with hyphens do not work on Posix platforms.
* I am planning to add support for modifiers, and with modifiers it will look even less related to RFC 1766.

Author: Well, I think that it is worth to add a separate section for locale names, with all specifications and links.

Reviewer 2: Perhaps you're right. The RFCs use hyphens, but glibc rejects such language codes, e.g. "de_DE.ISO8859-15@euro" works, but "de-DE.ISO8859-15@euro" is unsupported.

   determined.
   The "C" locale is represented as ``(None, None)``.

   .. deprecated-removed:: 3.11 3.15

.. function:: getlocale(category=LC_CTYPE)

   Returns the current setting for the given locale category as a tuple containing

>    language code and encoding. *category* may be one of the :const:`!LC_\*` values
>    except :const:`LC_ALL`.  It defaults to :const:`LC_CTYPE`.

Reviewer: ```suggestion
   the language code and encoding. *category* may be one of the :const:`!LC_\*`
   values except :const:`LC_ALL`.  It defaults to :const:`LC_CTYPE`.
```
Little grammar thing.

>    The format of the language code is platform depended, but on Posix

Reviewer: ```suggestion
   The format of the language code is platform dependent, but on Posix
```

>    platforms it usually looks like ``language[_territory]``.

Reviewer 2: Same as above: "looks like language[_territory] (see :rfc:1766)."

part of a character class such as letter or whitespace.

.. _locale_name:

Locale names
------------

The format of the locale name is platform dependent, and the set of supported
locales can depend on the system configuration.

> On Posix platforms, it usually has the format

Reviewer: ```suggestion
On Posix platforms, it usually has the format:
```

.. productionlist:: locale_name
   : language ["_" territory] ["." charset] ["@" modifier]

where *language* is a two- or three-letter language code from `ISO 639`_,
*territory* is a two-letter country or region code from ISO 3166,
*charset* is a locale encoding, and *modifier* is a script name,
a language subtag, a sort order identifier, or other locale modifier

> (e.g. "latin", "valencia", "stroke" and "euro").

Reviewer: ```suggestion
(for example, "latin", "valencia", "stroke" and "euro").
```
Style guide: Use simple language

On Windows, several formats are supported.
A subset of `IETF BCP 47`_ tags:

.. productionlist:: locale_name
   : language ["-" script] ["-" territory] ["." charset]
   : language ["-" script] "-" territory "-" modifier

> where *language* and *territory* has the same meaning as in Posix,

Reviewer: ```suggestion
where *language* and *territory* have the same meaning as in Posix,
```

*script* is a four-letter script code from `ISO 15924`_,
and *modifier* is a language subtag, a sort order identifier

> or custom modifier (e.g. "valencia", "stroke" or "x-python").

Reviewer: ```suggestion
or custom modifier (for example, "valencia", "stroke" or "x-python").
```

> Both hyphen ("``-``") and underscore ("``_``") separators are supported.

Reviewer: ```suggestion
Both hyphen (``'-'``) and underscore (``'_'``) separators are supported.
```
This is the more common (in `/Doc`), and IMO cleaner way.

Only UTF-8 encoding is allowed for BCP 47 tags.

> Windows supports also locale names in the format

Reviewer: ```suggestion
Windows also supports locale names in the format:
```

.. productionlist:: locale_name
   : language ["_" territory] ["." charset]

> where *language* and *territory* are long names, such as "English" and

Reviewer: ```suggestion
where *language* and *territory* are full names, such as "English" and
```

This is more clear IMO

> "United States", and *charset* is either a code page number (e.g. "1252")

Reviewer: ```suggestion
"United States", and *charset* is either a code page number (for example, "1252")
```

or UTF-8.
Only the underscore separator is supported in this format.

The "C" locale is supported on all platforms.

.. _ISO 639: https://www.iso.org/iso-639-language-code
.. _IETF BCP 47: https://www.rfc-editor.org/info/bcp47
.. _ISO 15924: https://www.unicode.org/iso15924/

> .. https://pubs.opengroup.org/onlinepubs/9799919799/basedefs/V1_chap08.html#tag_08_02

Reviewer: This could be a `.. seealso::`?

Author: They are just comments. I thought about adding links or notes, but I couldn't find a good anchor for that.
Anyway, they are useful even as comments, so we can check the future changes.

Thank you for your suggestions.
